File: src/main/scala/entities/Entities.scala
package entities

object Entities {
    case class Role(name: String, extend: Option[Role], min: Int, max: Int)
    case class Group(name: String, min: Int, max: Int, subgroups: List[Group], roles: List[Role], constraints: List[Constraint])

    trait Constraint
    case class Cardinality(min: Int, max: Int, subjectType: String, subject: String) extends Constraint
    case class Compatibility(from: String, to: String, scope: String, extendsSubgroups: Boolean, biDir: Boolean) extends Constraint

    case class Structural(roles: List[Role], groups: List[Group])

    case class Specification(structural: Structural)

    /*object Role {
        def apply(name: String, extend: Option[Role] = None, min: Int, max: Int): Role =
            Role(new ObjectId(), name, extend, min, max)
    }*/

    /*object Cardinality {
        def apply(min: Int, max: Int, subjectType: String, subject: String): Cardinality =
            Cardinality(new ObjectId(), min, max, subjectType, subject)
    }*/

    /*object Compatibility {
        def apply(from: String, to: String, scope: String, extendsSubgroups: Boolean, biDir: Boolean): Compatibility =
            Compatibility(new ObjectId(), from, to, scope, extendsSubgroups, biDir)
    }*/
}

File: src/main/kotlin/entities/Entities.kt
package entities

import org.bson.codecs.pojo.annotations.BsonId
import org.litote.kmongo.Id

object Entities {
    data class Role(val name: String, val extend: Role?, val min: Int, val max: Int)
    data class Group(
        val name: String,
        val min: Int,
        val max: Int,
        val subgroups: List<Group>,
        val roles: List<Role>,
        val cardinalities: List<Cardinality>,
        val compatibilities: List<Compatibility>
    )

    sealed interface Constraint
    data class Cardinality(val min: Int, val max: Int, val subjectType: String, val subject: String) : Constraint
    data class Compatibility(val from: String, val to: String, val scope: String, val extendsSubgroups: Boolean, val biDir: Boolean) : Constraint

    data class Structural(val roles: List<Role>, val groups: List<Group>)
    data class Specification(@BsonId val id: Id<Specification>, val structural: Structural)
}
